### PR TITLE
Add support for XEP-0107: User Mood

### PR DESCRIPTION
**Pull request by vanitasvitae**
This code is kinda untested (integration test did not yet run through successfully), as my raspi just died :(

I had to get a little distraction from OpenPGP, so this is the result :)

* [ ] Fix integration test (PEP updates are not arriving on my machine, though they do with @Flowdalic )
* [x] Add issue key to commit message

### CODE REVIEW

File: smack-extensions/src/main/java/org/jivesoftware/smackx/mood/MoodListener.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ */
+package org.jivesoftware.smackx.mood;
+
+import org.jivesoftware.smackx.mood.element.MoodElement;
+
+import org.jxmpp.jid.BareJid;
+
+public interface MoodListener {
+
+    void onMoodUpdated(BareJid jid, MoodElement moodElement);
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/mood/MoodListener.java]: Such listeners should also provide a reference to the Stanza which caused their invocation. In this case, add a third `Message` argument.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/mood/MoodManager.java [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    private static final Map<XMPPConnection, MoodManager> INSTANCES = new WeakHashMap<>();
+
+    private final Set<MoodListener> moodListeners = new HashSet<>();
+
+    private MoodManager(XMPPConnection connection) {
+        super(connection);
+        PEPManager.getInstanceFor(connection).addPEPListener(new PEPListener() {
+            @Override
+            public void eventReceived(final EntityBareJid from, final EventElement event, Message message) {
+                if (MOOD_NODE.equals(event.getEvent().getNode())) {
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/mood/MoodManager.java]: Try to avoid one level of indentation using `(!MOOD_NODE…`.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/mood/MoodManager.java [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+                    final BareJid contact = from.asBareJid();
+                    Async.go(new Runnable() {
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/mood/MoodManager.java]: This will lose the order of the mood events. You want to use Smack's new `AsyncButOrdered` tool here.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/mood/MoodManager.java [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+            SmackException.NotConnectedException, XMPPException.XMPPErrorException, PubSubException.NotALeafNodeException {
+        MoodElement element = buildMood(null, null, null);
+        publishMood(element);
+    }
+
+    private void publishMood(MoodElement moodElement)
+            throws SmackException.NotLoggedInException, InterruptedException, PubSubException.NotALeafNodeException,
+            XMPPException.XMPPErrorException, SmackException.NotConnectedException, SmackException.NoResponseException {
+
+        if (!connection().isAuthenticated()) {
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/mood/MoodManager.java]: The `Manager` class provides `getAuthenticatedConnectionOrThrow()` for this case.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/mood/MoodManager.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+            throw new SmackException.NotLoggedInException();
+        }
+        PubSubManager pubSubManager = PubSubManager.getInstance(connection(), connection().getUser().asBareJid());
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/mood/MoodManager.java]: A `Manager`'s `getInstanceFor()` method is expensive, why not have `MoodManager` hold a reference to `PubSubManager` in  a private field?

File: smack-extensions/src/main/java/org/jivesoftware/smackx/mood/MoodManager.java [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        return new MoodElement(
+                new MoodElement.MoodSubjectElement(mood, concretisation),
+                text);
+    }
+
+    public void addMoodToMessage(Message message, Mood mood) {
+        addMoodToMessage(message, mood, null);
+    }
+
+    public void addMoodToMessage(Message message, Mood mood, MoodConcretisation concretisation) {
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/mood/MoodManager.java]: Should be `static`

File: smack-extensions/src/main/java/org/jivesoftware/smackx/mood/MoodManager.java [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+        node.publish(new PayloadItem<>(moodElement));
+    }
+
+    private MoodElement buildMood(Mood mood, MoodConcretisation concretisation, String text) {
+        return new MoodElement(
+                new MoodElement.MoodSubjectElement(mood, concretisation),
+                text);
+    }
+
+    public void addMoodToMessage(Message message, Mood mood) {
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/mood/MoodManager.java]: Should be `static`

File: smack-extensions/src/main/java/org/jivesoftware/smackx/mood/MoodManager.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+        addMoodToMessage(message, mood, null);
+    }
+
+    public void addMoodToMessage(Message message, Mood mood, MoodConcretisation concretisation) {
+        MoodElement element = buildMood(mood, concretisation, null);
+        message.addExtension(element);
+    }
+
+    public void addMoodListener(MoodListener listener) {
```
Comment (Flowdalic) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/mood/MoodManager.java]: The methods to add or remove a listener are usually thread-safe in Smack. This is not the case here.

**Comment by vanitasvitae** [comment]
Will add the issue key as soon as the integration test is fixed.

File: documentation/extensions/index.md [hunk 9]
```diff
@@ -54,8 +54,9 @@ Smack Extensions and currently supported XEPs of smack-extensions
...
 | Stream Initation                            | [XEP-0095](https://xmpp.org/extensions/xep-0095.html)  | n/a       | Initiating a data stream between any two XMPP entities. |
 | [SI File Transfer](filetransfer.md)         | [XEP-0096](https://xmpp.org/extensions/xep-0096.html)  | n/a       | Transfer files between two users over XMPP. |
+| User Mood                                   | [XEP-0107](https://xmpp.org/extensions/xep-0107.html)  | 1.2.1     | Communicate the users current mood. |
 | [Entity Capabilities](caps.md)              | [XEP-0115](https://xmpp.org/extensions/xep-0115.html)  | n/a       | Broadcasting and dynamic discovery of entity capabilities. |
-| [Jingle](jingle.html)                       | [XEP-0116](http://xmpp.org/extensions/xep-0166.html)   | n/a       | Initiate and manage sessions between two XMPP entities. |
+| [Jingle](jingle.html)                       | [XEP-0116](http://xmpp.org/extensions/xep-0166.html)  | n/a       | Initiate and manage sessions between two XMPP entities. |
```
Comment (Flowdalic) [inline documentation/extensions/index.md]: Unrelated whitespace change, please add an extra commit

File: smack-integration-test/src/main/java/org/jivesoftware/smackx/mood/MoodIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+    @SmackIntegrationTest
+    public void test() throws Exception {
+        MoodManager mm1 = MoodManager.getInstanceFor(conOne);
+        MoodManager mm2 = MoodManager.getInstanceFor(conTwo);
+
+        Roster r1 = Roster.getInstanceFor(conOne);
+        Roster r2 = Roster.getInstanceFor(conTwo);
+
+        syncSubscribe(conOne, conTwo, "two");
```
Comment (Flowdalic) [inline smack-integration-test/src/main/java/org/jivesoftware/smackx/mood/MoodIntegrationTest.java]: Please use `RosterIntegrationTest.ensureBothAccountsAreSubscribedToEachOther()` instead.

File: smack-integration-test/src/main/java/org/jivesoftware/smackx/mood/MoodIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+public class MoodIntegrationTest extends AbstractSmackIntegrationTest {
+
+    public MoodIntegrationTest(SmackIntegrationTestEnvironment environment) {
+        super(environment);
+    }
+
+    @SmackIntegrationTest
+    public void test() throws Exception {
+        MoodManager mm1 = MoodManager.getInstanceFor(conOne);
```
Comment (Flowdalic) [inline smack-integration-test/src/main/java/org/jivesoftware/smackx/mood/MoodIntegrationTest.java]: `moodManager(1|2)` should be final fields of the test and set in the constructor.

File: smack-integration-test/src/main/java/org/jivesoftware/smackx/mood/MoodIntegrationTest.java [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+                if (moodElement.getMood() == Mood.satisfied) {
+                    moodReceived.signal();
+                }
+            }
+        });
+
+        mm1.setMood(Mood.satisfied);
+
+        moodReceived.waitForResult(timeout);
+        r1.removeEntry(r1.getEntry(conTwo.getUser().asBareJid()));
```
Comment (Flowdalic) [inline smack-integration-test/src/main/java/org/jivesoftware/smackx/mood/MoodIntegrationTest.java]: Those should be wrapped in a `finally` block, or even better, in a `@AfterClass` annotated method

File: smack-integration-test/src/main/java/org/jivesoftware/smackx/mood/MoodIntegrationTest.java (hunk 10 above)
Comment (Flowdalic) [reply smack-integration-test/src/main/java/org/jivesoftware/smackx/mood/MoodIntegrationTest.java]: Yeah, I'm not sure if the API is where it should be, as it can be considered surprisingly to find stuff like that in an `IntegrationTest`. We should probably move it into a `IntegrationTestRosterUtil` somewhere in smack-integration-test. I've added this on my TODO list after this PR got merged.
